### PR TITLE
waivers: functional test fixture and eval-time skipping

### PR DESCRIPTION
**Pull request by clintoncwolfe**
## Description

This is an implementation of #3845/#4414, which uses inputs to act as if `only_if` had been defined for a control. The waiver rules are applied, and the control is conditionally skipped using the only_if facility.

This PR includes a test fixture profile, which has a control and corresponding Input rotating through 11 cases of the various ways waivers may be used.

There is also a passing functional test.  It really does test that the controls are skipped when waivered; however there are several no-op assertions in the functional test file that are just placeholders for  Part 3.

There is an input file included. I made up the field names - they can certainly be adjusted as can the values.

There are placeholders in the functional test file for assertion methods for detecting3845 waiver annotations in the JSON report.

I went with a functional test after staring at the unit tests for the profile and control_eval_context unit tests. That's where the internal unit tests belong, but that is really opaque code. I was able to get a functional test up and running, and start tracing things that way much more easily. Happy to circle back later on that.

## Review Notes

Several things to discuss:

* ~~The waiver rules are currently applied BEFORE evaulating the contents of the control. I believe that means that any real `only_if` block in the profile source will have higher precedence.~~ UPDATE: the waiver rules now apply after the only_if.

* the ability to "slug" the control IDs (replacing spaces, etc) is not implemented here but needs to be implemented somewhere - not sure if here is the right place. @zenspider?

* ~~As implemented, some of the waiver logic (such as expiration date comparision) might need to be re-created in step 3 #4417. We may want to look into how to avoid that.~~ UPDATE: all logic happens here in step 2 now, and it records information in the run_data for step 3 to display, hopefully with only presentation (not business) logic.

## Related Issue

#3845 , #4414 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New content (non-breaking change)
- [ ] Breaking change (a content change which would break existing functionality or processes)

## Checklist:

- [ ] I have read the **CONTRIBUTING** document.

### CODE REVIEW

File: lib/inspec/rule.rb [hunk 1]
```diff
@@ -42,11 +43,16 @@ def initialize(id, profile_id, opts, &block)
       @__rule_id = id
       @__profile_id = profile_id
       @__checks = []
-      @__skip_rule = {}
+      @__skip_rule = {} # { result: true, message: "Why" }
```
Comment (zenspider) [inline lib/inspec/rule.rb]: By piggybacking here, does that mean that "last one wins" and thus, we might lose the information of the waiver? I don't know how `only_if` works, but I assume it'll re-set some of these fields and we'll lose information. We might want to prefix fields in this PR with `waiver_` or something similar.

IOW, how does automate differentiate between a normal only_if triggering and a waiver triggering?

File: lib/inspec/rule.rb [hunk 2]
```diff
@@ -267,6 +275,45 @@ def __add_check(describe_or_expect, values, block)
       @__checks.push([describe_or_expect, values, block])
     end
 
+    # Look for an input with a matching ID, and if found, apply waiver
+    # skipping logic. Basically, if we have a current waiver, and it says
+    # to skip, we'll replace all the checks with a dummy check (same as
+    # only_if mechanism)
+    # Double underscore: not intended to be called as part of the DSL
```
Comment (zenspider) [inline lib/inspec/rule.rb:285]: `private`?

File: lib/inspec/rule.rb [hunk 3]
```diff
@@ -42,11 +43,16 @@ def initialize(id, profile_id, opts, &block)
...
       @__checks = []
-      @__skip_rule = {}
+      @__skip_rule = {} # { result: true, message: "Why" }
       @__merge_count = 0
       @__merge_changes = []
       @__skip_only_if_eval = opts[:skip_only_if_eval]
 
+      # TODO: Note - if waivers are applied before the DSL instance
+      # eval, then the contents of the control block (including only_if)
+      # will override the waiver. Is that wanted? Or should waivers always win?
```
Comment (zenspider) [inline lib/inspec/rule.rb]: I think this is what I was trying to say above. We need waivers to win OR we need to distinguish between the two via the same data structure.

File: test/functional/waivers_test.rb [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+describe "waivers" do
+  include FunctionalHelper
+  let(:waivers_profiles_path) { "#{profile_path}/waivers" }
+
+  def assert_test_outcome(expected, result_json)
+    assert_equal expected, result_json["status"]
+  end
+
+  def assert_waiver_annotation(result_json)
+    # TODO - test JSON for waiver annotation
```
Comment (zenspider) [inline test/functional/waivers_test.rb:15]: these should call `flunk` to make it obvious that they're not done.

File: test/functional/waivers_test.rb [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+  def refute_waiver_annotation(result_json)
+    # TODO - test JSON for waiver annotation
+    # Don't suppose we get this for free by defining assert_waiver_annotation ...
+  end
+
+  describe "a fully pre-slugged control file" do
+    it "has all of the expected outcomes" do
+      cmd = "exec #{waivers_profiles_path}/basic --input-file #{waivers_profiles_path}/basic/files/waivers.yaml"
+      run_result = run_inspec_process(cmd, json: true)
+      controls_by_id = run_result.payload.json["profiles"][0]["controls"].map { |c| [c["id"], c] }.to_h
```
Comment (zenspider) [inline test/functional/waivers_test.rb]: I'm curious why you use `dig` for a single level above but don't use it here for 3 levels here?

File: test/functional/waivers_test.rb [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        assert_test_outcome "passed", result
+        refute_waiver_annotation result
+      end
+
+      [
+        "02_not_waivered_fails",
+        "07_waivered_expiry_in_past_not_skipped_fails", # Should this give a special waiver expired message?
+      ].each do |control_id|
+        result = controls_by_id[control_id]["results"][0]
+        assert_test_outcome "failed", result
```
Comment (zenspider) [inline test/functional/waivers_test.rb]: The repetition here with `result =` and the `assert_*` between all of these `each` blocks hints that `control_id` should be the thing passed in:

```ruby
let(:controls_by_id) { run_result.payload.json.dig("profiles", 0, "controls").map { |c| [c["id"], c] }.to_h }
# ...
assert_test_outcome "failed", control_id
```

File: test/functional/waivers_test.rb [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+        assert_waiver_annotation result
+      end
+
+    end
+  end
+
+  # describe "an inherited profile"
+  # describe "a profile whose control ids require transformation"
```
Comment (zenspider) [inline test/functional/waivers_test.rb:69]: These could be real describes with an `it` that has a `skip` so we have something in our output.

File: test/unit/mock/profiles/waivers/basic/controls/basic.rb [hunk 8]
```diff
@@ -0,0 +1,43 @@
+control "01_not_waivered_passes" do
+  describe(true) { it { should eq true } }
```
Comment (zenspider) [inline test/unit/mock/profiles/waivers/basic/controls/basic.rb:2]: I find this extra verbage confusing tbh... The true/false values are not important and adding noise. The nominal cases could be something simple like:

```ruby
describe(:passes) { it { :passes } }
```
and the waived case could be
```ruby
describe(:fails) { it { fail "don't run me" } }
```

making it so you don't have to think about the positive case at all because it both tells you that it passes but also because it essentially does nothing. I don't have to `eval` it in my head. Likewise, the negative case literally tells me that it fails and then does exactly that. Again, no `eval` in my head.

File: test/unit/mock/profiles/waivers/basic/controls/basic.rb [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+control "09_waivered_expiry_in_future_not_skipped_passes" do
+  describe(true) { it { should eq true } }
+end
+
+control "10_waivered_expiry_in_future_not_skipped_fails" do
+  describe(true) { it { should eq false } }
+end
+
+control "11_waivered_expiry_in_future_skipped" do
+  describe(true) { it { should eq true } }
```
Comment (zenspider) [inline test/unit/mock/profiles/waivers/basic/controls/basic.rb:42]: This can lead to a false positive if we don't implement anything because this dosn't fail.

File: test/unit/mock/profiles/waivers/basic/files/waivers.yaml [hunk 10]
```diff
@@ -0,0 +1,44 @@
+waiver_03_waivered_no_expiry_not_skipped_passes:
+  expiration_date: never
+  justification: Sound reasoning
+  skip: no
```
Comment (zenspider) [inline test/unit/mock/profiles/waivers/basic/files/waivers.yaml:3]: I think this file is both covering too much and missing some steps. The edge cases as I see them:

* expiration_date: nil (we said it was optional), past, future
* skip: yes, no
* test result: passing or failing

and so we _could_ have 12 waiver combinations... but in reality we don't need to. For example, past versus future doesn't need to test every combination of skip or test result, just that it does invoke properly when in the future and that past acts the same as nil. See waiver `09` and `10` for what I think of as redundant tests, and expiration as `nil` as a missing test.

File: lib/inspec/rule.rb [hunk 11]
```diff
@@ -267,6 +275,45 @@ def __add_check(describe_or_expect, values, block)
       @__checks.push([describe_or_expect, values, block])
     end
 
+    # Look for an input with a matching ID, and if found, apply waiver
+    # skipping logic. Basically, if we have a current waiver, and it says
+    # to skip, we'll replace all the checks with a dummy check (same as
+    # only_if mechanism)
+    # Double underscore: not intended to be called as part of the DSL
+    def __apply_waivers
+      input_name = "waiver_#{@__rule_id}" # TODO: control ID slugging
```
Comment (zenspider) [inline lib/inspec/rule.rb]: The design spec we came up with only used the ID and didn't prefix with `waiver_`. I thought we were keying in on the example document from the design meeting and would just be looking for "skip":

https://github.com/inspec/inspec/issues/4409#issuecomment-525049831

Is there a reason for this?

File: lib/inspec/rule.rb [hunk 12]
```diff
@@ -267,6 +275,45 @@ def __add_check(describe_or_expect, values, block)
...
+      registry = Inspec::InputRegistry.instance
+      input = registry.inputs_by_profile[@__profile_id].dig(input_name)
```
Comment (zenspider) [inline lib/inspec/rule.rb]: I don't understand using `dig` this way. Wouldn't using it for both levels make more sense?

```ruby
input = registry.inputs_by_profile.dig(@__profile_id, input_name)
```

File: lib/inspec/rule.rb [hunk 13]
```diff
@@ -267,6 +275,45 @@ def __add_check(describe_or_expect, values, block)
...
+      return unless input
+
+      input = input.value
```
Comment (zenspider) [inline lib/inspec/rule.rb]: This probably needs a comment explaining the datastructure involved. You did a `dig` above to `input_name`, so don't you have the value?

And I'm very confused digging into this and trying to parse:
https://github.com/inspec/inspec/blob/435747a2d2b4a52b7044eda65133bdbb19121acf/lib/inspec/input_registry.rb#L24

File: lib/inspec/rule.rb [hunk 14]
```diff
@@ -267,6 +275,45 @@ def __add_check(describe_or_expect, values, block)
...
+
+      # Waivers should have a hash value with keys including skip and
+      # expiration_date. We only care here if it has a skip key and it
```
Comment (zenspider) [inline lib/inspec/rule.rb:302]: The design stated that expiration date was optional.

File: lib/inspec/rule.rb [hunk 15]
```diff
@@ -267,6 +275,45 @@ def __add_check(describe_or_expect, values, block)
...
+      # is yes-like, since all non-skipped waiver operations are handled
+      # during reporting phase.
+      return unless input.key?("skip")
+      return unless input["skip"].to_s.match(/y|yes|true/i)
```
Comment (zenspider) [inline lib/inspec/rule.rb]: We specified this to be a boolean and should treat it as such... This could just be:
```ruby
return unless input["skip"]
```

BUT! a true `skip` meant "don't run this" and I would think we'd need to attach the waiver data somewhere in the case of a non-skipped-but-waived control, no?

File: lib/inspec/rule.rb [hunk 16]
```diff
@@ -267,6 +275,45 @@ def __add_check(describe_or_expect, values, block)
...
+
+      # OK, the intent is to skip. Does it have an expiration date, and
+      # if so, is it in the future?
+      expiry = input["expiration_date"]
+      if expiry
+        if expiry.is_a?(Date)
+          # It appears that yaml.rb automagically parses dates for us
+          return if expiry < Date.today # If the waiver expired, return - no skip applied
+        elsif expiry.match(/never/i) # any other values?
```
Comment (zenspider) [inline lib/inspec/rule.rb]: This was not in the design.

File: lib/inspec/rule.rb [hunk 17]
```diff
@@ -267,6 +275,45 @@ def __add_check(describe_or_expect, values, block)
...
+      return unless input.key?("skip")
+      return unless input["skip"].to_s.match(/y|yes|true/i)
+
+      # OK, the intent is to skip. Does it have an expiration date, and
+      # if so, is it in the future?
+      expiry = input["expiration_date"]
+      if expiry
+        if expiry.is_a?(Date)
+          # It appears that yaml.rb automagically parses dates for us
+          return if expiry < Date.today # If the waiver expired, return - no skip applied
```
Comment (zenspider) [inline lib/inspec/rule.rb]: I guess we never specified whether `expiration_date` meant "through" or "until"... so this might need to be a `<=`?

File: lib/inspec/rule.rb [hunk 18]
```diff
@@ -267,6 +275,45 @@ def __add_check(describe_or_expect, values, block)
...
+        elsif expiry.match(/never/i) # any other values?
+          # Do nothing, fall through
+        else
+          raise Inspec::Exceptions::ResourceFailed, "Unable to parse waiver expiration date '#{expiry}' for control #{@__rule_id}"
```
Comment (zenspider) [inline lib/inspec/rule.rb]: We parsed it (valid yaml)... it just has invalid data.

File: lib/inspec/rule.rb [hunk 19]
```diff
@@ -267,6 +275,45 @@ def __add_check(describe_or_expect, values, block)
...
+        end
+      end
+
+      # OK, apply a skip.
+      @__skip_rule[:result] = true
+      @__skip_rule[:message] = "Skipped due to waiver: #{input["justification"] || "(no reason given)"}"
```
Comment (zenspider) [inline lib/inspec/rule.rb]: I might be misunderstanding the point of `@__skip_rule`, but to me this should be passing through the expiration date and whether it was skipped?

...or does the backend need to re-process all of the same input/waiver data?

The answer to this question determines the overall implementation of step 3.

Comment (zenspider) [reply lib/inspec/rule.rb]: Also, since the tests are laid out for step 3... should @miah be working _on top_ of your branch?

**Review by zenspider — CHANGES_REQUESTED** [review]
Applied a critical eye to this one since it determines a lot for step 3 (and step 1 to a lesser degree).

File: lib/inspec/rule.rb (hunk 2 above)
Comment (clintoncwolfe) [reply lib/inspec/rule.rb:285]: The body of each control is instance eval'd against an instance of this class - so `private` won't stop someone from writing profile code that tinkers with this stuff. (though there is a `private` declaration later in the file.)

File: test/functional/waivers_test.rb (hunk 5 above)
Comment (clintoncwolfe) [reply test/functional/waivers_test.rb]: I think of it more for cases when I don't know if an intermediate is nil, vs avoiding a long chain of references. It's prettier as a dig here, certainly.

File: test/unit/mock/profiles/waivers/basic/controls/basic.rb (hunk 8 above)
Comment (clintoncwolfe) [reply test/unit/mock/profiles/waivers/basic/controls/basic.rb:2]: Well, I think this comes from different perspectives. I'm not someone with years of Ruby testing experience (so my brain doesn't see the same "labor pathways" as yours), but I am someone with some InSpec exposure; so my goal is to make the test fixture look like a minimal inspec profile. I would not want someone - like a new contributor - to open a simple test file and get hung up on unfamiliar code; it should just look like InSpec, unless we have a strong reason for it not to look like InSpec. And InSpec usually has `should op value` stanzas. So, that's why I provide them.

File: test/unit/mock/profiles/waivers/basic/controls/basic.rb (hunk 9 above)
Comment (clintoncwolfe) [reply test/unit/mock/profiles/waivers/basic/controls/basic.rb:42]: This is the test fixture, not the test. The test is to verify that this should skip, not pass. If it passes, that's a defect, and our test should detect that.

We can add an additional case that would fail if not skipped, if you like.

File: test/unit/mock/profiles/waivers/basic/files/waivers.yaml (hunk 10 above)
Comment (clintoncwolfe) [reply test/unit/mock/profiles/waivers/basic/files/waivers.yaml:3]: I think we should probably draw up a formal IO table and show it to Product to verify behavior.  In the meantime I've accepted the note about nil expiration and am adding test cases to cover.

I don't see any harm in redundancy in test cases - I'm much more concerned about thoroughness, and seeing every case covered.

File: lib/inspec/rule.rb (hunk 14 above)
Comment (clintoncwolfe) [reply lib/inspec/rule.rb:302]: I'll alter behavior as follows:
* if absent, no expiration date (never expires)
* if present, must be a valid parseable date
* no other values valid (no special handling of "never", etc)

File: lib/inspec/rule.rb (hunk 18 above)
Comment (clintoncwolfe) [reply lib/inspec/rule.rb]: No, it would have been parsed twice - the YAML parser, then again as a Date. I'm saying the second parsing failed.

File: lib/inspec/rule.rb (hunk 11 above)
Comment (clintoncwolfe) [reply lib/inspec/rule.rb]: No, just headcanon. Removed.

File: lib/inspec/rule.rb (hunk 12 above)
Comment (clintoncwolfe) [reply lib/inspec/rule.rb]: Yes. I was concerned because the registry delegates part of that - but yes, here, that's clearly fine.

File: lib/inspec/rule.rb (hunk 13 above)
Comment (clintoncwolfe) [reply lib/inspec/rule.rb]: I've renamed the variable and added a clarifying comment.

File: lib/inspec/rule.rb (hunk 15 above)
Comment (clintoncwolfe) [reply lib/inspec/rule.rb]: I've been assuming that the job of phase 2 is to handle skipping - alterations to the runtime execution plan - and phase three is to handle reporting.  Both phase 2 and 3 can look to the InputRegistry as the sole source of truth for waiver data. 

I think if we were to stash that data somewhere now, I'd feel like we were just making another copy of it for unclear reasons. Unless we want to avoid scanning the entire list of controls during reporting?

File: lib/inspec/rule.rb (hunk 17 above)
Comment (clintoncwolfe) [reply lib/inspec/rule.rb]: Your call :-)

File: lib/inspec/rule.rb (hunk 19 above)
Comment (clintoncwolfe) [reply lib/inspec/rule.rb]: As I understand it, `@__skip_rule` is a data structure that determines if the control has been marked to be skipped. If `@__skip_rule[:result]` is true, then the runtime will "skip" the tests (by replacing all of them with one dummy `os` check, see `prepare_checks`).  `only_if` uses the same mechanism.

I'm all for altering the fields in that structure to make it more flexible - right now it adds bad messaging, assuming all skips are "Only if." We could also invent a new mechanism.

All of this here, though, only affects the textual messaging associated with the control. For step three, we're wanting new separate fields in the JSON output. And I think - with some code deduplication - it should be the same sort of logic we see in step 2 - read from the Input, then consider if it was expired, was it to be skipped, etc.  I'm not sure how much of the Rule datastructure we have access to during the reporting phase (if any) - clear that's something to look into ASAP.

Comment (clintoncwolfe) [reply lib/inspec/rule.rb]: OK, I've located where we'd stash such data in the RSpec example metadata - we have an opportunity here - https://github.com/inspec/inspec/blob/d1ffce727828853222b80ae018d969b2f6c2bc74/lib/inspec/runner_rspec.rb#L166 - however data structure there does not propagate to the run_data hash passed to the reporter structure. I'm still looking at that.

Comment (clintoncwolfe) [reply lib/inspec/rule.rb]: I found where the RSpec metadata may be propagated to the reporter Hash data structure - here - https://github.com/inspec/inspec/blob/master/lib/inspec/formatters/base.rb#L152 .

At this point we could concentrate all waiver logic in step 2, and only do display in step 3.

File: lib/inspec/rule.rb (hunk 1 above)
Comment (clintoncwolfe) [reply lib/inspec/rule.rb]: I've updated the PR to now distinguish between the two types of skip; moved the apply_waivers call to ensure it takes precedence; and added functional tests to verify.

File: lib/inspec/rule.rb (hunk 15 above)
Comment (clintoncwolfe) [reply lib/inspec/rule.rb]: Having slept on it and done some more code reading, I think doing the logic once here in step 2, then attaching the data somewhere (the control metadata) seems the most sensible. I've done so on d118ed62.

File: lib/inspec/rule.rb (hunk 19 above)
Comment (clintoncwolfe) [reply lib/inspec/rule.rb]: I've cleaned up the way that the messaging is handled, and placed all of the business logic around waivers in step 2. AFAIK in step 3 now we just need to display fields - I've updated the design doc with notes on what should be available in the `run_data` structure.

File: lib/inspec/rule.rb (hunk 15 above)
Comment (zenspider) [reply lib/inspec/rule.rb]: The above is still trying to do the job that YAML already does:

```
9982 % ruby -ryaml -e 'p YAML.load("---\nkey: yes")'
{"key"=>true}
```

File: lib/inspec/rule.rb (hunk 17 above)
Comment (zenspider) [reply lib/inspec/rule.rb]: I don't really care as long as it is documented. I would probably assume that "expiration_date" means "on", thus "<".

File: lib/inspec/rule.rb (hunk 18 above)
Comment (zenspider) [reply lib/inspec/rule.rb]: yaml handles dates:

```
9985 % ruby -ryaml -e 'p YAML.load("---\nkey: 2019-09-01")'
{"key"=>#<Date: 2019-09-01 ((2458728j,0s,0n),+0s,2299161j)>}
```

I'm saying, either the yaml parses or it doesn't... the rest is about invalid data, not parsing.

File: test/unit/mock/profiles/waivers/basic/files/waivers.yaml (hunk 10 above)
Comment (zenspider) [reply test/unit/mock/profiles/waivers/basic/files/waivers.yaml:3]: **M**VP

**Review by zenspider — APPROVED** [review]
I think there's still some stuff going on that doesn't need to (eg stringifying and re-parsing already parsed yaml data)... but this should be complete enough to work on phase 3

File: lib/inspec/rule.rb [hunk 20]
```diff
@@ -42,7 +44,7 @@ def initialize(id, profile_id, opts, &block)
       @__rule_id = id
       @__profile_id = profile_id
       @__checks = []
-      @__skip_rule = {}
+      @__skip_rule = {} # { result: true, message: "Why", type: [:only_if, :waiver] }
```
Comment (miah) [inline lib/inspec/rule.rb:47]: Should this be an real object? (not a blocker, just a concern)

File: lib/inspec/rule.rb [hunk 21]
```diff
@@ -206,14 +214,16 @@ def self.merge_changes(rule)
       rule.instance_variable_get(:@__merge_changes)
     end
 
+    # If a rule is marked to be skipped, this
+    # creates a dummay array of "checks" with a skip outcome
```
Comment (miah) [inline lib/inspec/rule.rb:218]: s/dummay/dummy

**Review by miah — APPROVED** [review]
